test/object-store/sync/flx_sync.cpp

```diff
@@ -694,6 +697,86 @@ TEST_CASE("flx: connect to PBS as FLX returns an error", "[sync][flx][app]") {
+        REQUIRE(token.timestamp);
+        REQUIRE(token.expires);
+        REQUIRE(token.timestamp < token.expires);
+        std::chrono::system_clock::time_point now = std::chrono::system_clock::now();
+        using namespace std::chrono_literals;
+        token.expires = std::chrono::system_clock::to_time_t(now - 30s);
+        REQUIRE(token.expired(now));
+    }
+
+    class HookedTransport : public SynchronousTestTransport {
+    public:
+        void send_request_to_server(Request&& request,
+                                    util::UniqueFunction<void(const Response&)>&& completion_block) override
+        {
+            if (request_hook) {
+                request_hook(request);
+            }
+            SynchronousTestTransport::send_request_to_server(std::move(request), [&](const Response& response) {
+                if (response_hook) {
+                    response_hook(request, const_cast<Response&>(response));
+                }
+                completion_block(response);
+            });
+        }
+        util::UniqueFunction<void(Request&, Response&)> response_hook;
+        util::UniqueFunction<void(Request&)> request_hook;
+    };
+
+    auto transport = std::make_shared<HookedTransport>();
+    FLXSyncTestHarness harness("flx_wait_access_token2", FLXSyncTestHarness::default_server_schema(), transport);
+
+    auto sync_mgr = harness.make_sync_manager();
+    auto creds = create_user_and_log_in(sync_mgr.app());
+    auto app = sync_mgr.app();
+    std::shared_ptr<SyncUser> user = app->current_user();
+    REQUIRE(user);
+    REQUIRE(!user->access_token_refresh_required());
+    // Set a bad access token, with an expired time. This will trigger a refresh initiated by the client.
+    user->update_access_token(encode_fake_jwt("fake_access_token", token.expires, token.timestamp));
+    REQUIRE(user->access_token_refresh_required());
+
+    bool seen_waiting_for_access_token = false;
+    // Commit a subcription set while there is no sync session.
+    // A session is created when the access token is refreshed.
+    transport->request_hook = [&](Request&) {
+        auto user = app->current_user();
+        REQUIRE(user);
+        for (auto& session : user->all_sessions()) {
+            if (session->state() == SyncSession::State::WaitingForAccessToken) {
+                REQUIRE(!seen_waiting_for_access_token);
+                seen_waiting_for_access_token = true;
+
+                REQUIRE(session->has_flx_subscription_store());
+                auto store = session->get_flx_subscription_store();
+                auto mut_subs = store->get_latest().make_mutable_copy();
+                std::move(mut_subs).commit();
+            }
+        }
+    };
+    SyncTestFile config(sync_mgr.app()->current_user(), harness.schema(), SyncConfig::FLXSyncEnabled{});
+    config.persist();
+    auto r = Realm::get_shared_realm(config);
+    REQUIRE(seen_waiting_for_access_token);
```

Comment:
How do we know this is guaranteed to be true immediately after calling get_shared_realm()?

Reply:
**get_shared_realm** triggers the token refresh that ends up with a server request. seen_waiting_for_access_token turns true when the server request is made. This test is very similar to https://github.com/realm/realm-core/blob/master/test/object-store/sync/app.cpp?#L2192